File: frc2018_5/src/main/java/ca/warp7/frc2018_5/io/DriveOutput.java
package ca.warp7.frc2018_5.io;

import ca.warp7.frc.next.Robot;
import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix.motorcontrol.can.VictorSPX;

import static ca.warp7.frc2018_5.constants.DriveConstants.kLeftDriftOffset;
import static ca.warp7.frc2018_5.constants.DriveConstants.kRightDriftOffset;
import static ca.warp7.frc2018_5.constants.RobotPins.*;

public class DriveOutput implements Robot.OutputSystem {

    private static DriveOutput instance;

    public static DriveOutput getInstance() {
        if (instance == null) instance = new DriveOutput();
        return instance;
    }

    private VictorSPX mLeftMaster;
    private VictorSPX mRightMaster;

    private double leftPercentOutput = 0;
    private double rightPercentOutput = 0;

    private DriveOutput() {
        mLeftMaster = new VictorSPX(kDriveLeftA);
        mRightMaster = new VictorSPX(kDriveRightA);

        VictorSPX leftFollower = new VictorSPX(kDriveLeftB);
        VictorSPX rightFollower = new VictorSPX(kDriveRightB);

        mRightMaster.setInverted(true);

        leftFollower.set(ControlMode.Follower, mLeftMaster.getDeviceID());
        rightFollower.set(ControlMode.Follower, mRightMaster.getDeviceID());
    }

    @Override
    public void onDisabled() {
        mLeftMaster.set(ControlMode.Disabled, 0);
        mRightMaster.set(ControlMode.Disabled, 0);
    }

    @Override
    public void onOutput() {
        mLeftMaster.set(ControlMode.PercentOutput, leftPercentOutput);
        mRightMaster.set(ControlMode.PercentOutput, rightPercentOutput);
    }

    public void setPercentOutput(double left, double right) {
        leftPercentOutput = Robot.limit(left, 1.0) * kLeftDriftOffset;
        rightPercentOutput = Robot.limit(right, 1.0) * kRightDriftOffset;
    }
}

File: frc2018_5/src/main/java/ca/warp7/frc2018_5/io/DriveOutput.kt
package ca.warp7.frc2018_5.io

import ca.warp7.frc.next.Robot
import ca.warp7.frc2018_5.constants.RobotPins.*
import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.can.VictorSPX
import edu.wpi.first.wpilibj.Solenoid

object DriveOutput : Robot.OutputSystem {

    private val mLeftMaster: VictorSPX = VictorSPX(kDriveLeftA)
    private val mRightMaster: VictorSPX = VictorSPX(kDriveRightA)
    private val mShifterSolenoid = Solenoid(kDriveShifterSolenoidPin)

    var leftPercentOutput = 0.0
    var rightPercentOutput = 0.0
    var solenoidOnForShifter: Boolean = false

    init {
        mRightMaster.inverted = true

        val leftFollower = VictorSPX(kDriveLeftB)
        val rightFollower = VictorSPX(kDriveRightB)
        leftFollower.set(ControlMode.Follower, mLeftMaster.deviceID.toDouble())
        rightFollower.set(ControlMode.Follower, mRightMaster.deviceID.toDouble())

        mShifterSolenoid.set(true)
        mShifterSolenoid.set(false)
    }

    @Synchronized
    override fun onDisabled() {
        mLeftMaster.set(ControlMode.Disabled, 0.0)
        mRightMaster.set(ControlMode.Disabled, 0.0)
    }

    @Synchronized
    override fun onOutput() {
        if (solenoidOnForShifter) {
            if (!mShifterSolenoid.get()) {
                mShifterSolenoid.set(true)
            }
        } else if (mShifterSolenoid.get()) {
            mShifterSolenoid.set(false)
        }

        mLeftMaster.set(ControlMode.PercentOutput, leftPercentOutput)
        mRightMaster.set(ControlMode.PercentOutput, rightPercentOutput)
    }
}
